### ISSUE TITLE
git rebase applies commits in wrong order

### ISSUE THREAD

**Issue by Mateiaru** [pcottle/learnGitBranching#660]
In level `advanced3` ("branch spaghetti"), execute the following commands, in order:

    git branch -f three C2
    git rebase -i one C5
    # and, in dialog, rearrange them C4, C3, C2, with C5 ommited
    git branch -f one
    git checkout two
    git cherry-pick C5

At this points, screen should look like this: 
![image](https://user-images.githubusercontent.com/8215371/77342254-96927c00-6d38-11ea-82af-49cba57de40e.png)
Next:

    git rebase two C2'

Result:
![image](https://user-images.githubusercontent.com/8215371/77342406-cf325580-6d38-11ea-841e-168ab3625138.png)

As you can see, the commits are applied in reverse order. Sometimes, they applied in a totally random order, like C2, C4, C3. To be sure, I have tried to do the exact same thing in real `git`. Setup, as seen from `gitk --all`, is:
![image](https://user-images.githubusercontent.com/8215371/77342702-4667e980-6d39-11ea-97b8-d55f236cf5df.png)
As you can see, the setup is exactly the same as in the first picture (except, of course, the commit message indicates the commit instead of its SHA, and that C2 is tagged for simplicity; also, I removed the prime (apostrophe) in commit names). Next, I do `git rebase two C2`. Output:
![image](https://user-images.githubusercontent.com/8215371/77343121-e02f9680-6d39-11ea-8c26-195a196837da.png)
Which puts the commits, in the correct order, on top of branch 'two'.

In conclusion, just like the title says, `git rebase` sometimes applies the commits in the wrong order.
A very simple and easy workaround is to do add `-i` to the rebase and then just hit confirm without changing anything; this will put them in the correct order.

**Comment by pcottle** [comment]
Thanks for dropping by @Mateiaru and pointing this out. this is indeed pretty odd, but I can't seem to repo in a normal sandbox. There's an easy one-click link to recreate the tree you're showing and rebase the commits:
https://learngitbranching.js.org/?NODEMO&command=gc;gc;gc;gc;git%20checkout%20-b%20one%20C1;%20git%20cherry-pick%20C4%20C3%20C2;%20git%20checkout%20-b%20two%20C1;%20git%20cherry-pick%20C5;%20git%20rebase%20two%20C2%27

I've ran it 4 or five times and it usually ends up in the correct order, but sometimes its something random like you pointed out! its probably some weird undefined ordering code deep within the rebase stuff. ill try to dive in here...

**Comment by pcottle** [comment]
Ah well here's one of the reasons why its difficult to repro -- it relies on the commits being created in the same second:

```
GitEngine.prototype.dateSortFunc = function(cA, cB) {
  var dateA = new Date(cA.get('createTime'));
  var dateB = new Date(cB.get('createTime'));
  if (dateA - dateB === 0) {
```

only if that codepath happens does it fall into ID comparing. let me dig further...

**Comment by pcottle** [comment]
<img width="782" alt="Screen Shot 2020-03-29 at 5 11 46 PM" src="https://user-images.githubusercontent.com/1135007/77864921-60de0f00-71e0-11ea-921d-5c6837eac695.png">
seems nicely consistent now :)

**Comment by hong4rc** [comment]
@pcottle It is fixed, isn't is?

Can you close all the fixed issue?

I am checking old issue, attempt to fix them.

**Comment by elodg** [comment]
This issue is still present.
![image](https://user-images.githubusercontent.com/7438625/80405568-7885e180-88cb-11ea-9eb4-b3ee671e4f99.png)

**Comment by pcottle** [comment]
I definitely fixed the random-order application which was a weird side effect with how we did sorting (and a race condition around time of commit creation). Things should be deterministic at least, so we at least improved that!

Now for the actual ordering itself, the weird thing here is that the cherrypick command is essentially "reversing" the stack by going to C4->C3->C2. Now for git, it respects the ordering of the stack during a rebase but for our purposes, we're basically sorting by autoincrement ID. So when the rebase happens, we assume the order is C2 -> C3 -> C4.

this different demo makes it seem less weird, since our cherrypick preserves the ordering that we want to rebase with:
https://learngitbranching.js.org/?NODEMO&command=gc;gc;gc;gc;git%20checkout%20-b%20one%20C1;%20git%20cherry-pick%20C2%20C3%20C4;%20git%20checkout%20-b%20two%20C1;%20git%20cherry-pick%20C5;%20git%20rebase%20two%20C4%27

Unfortunately "fixing" this would require a substantial refactor of how the git engine works, since the rebase step only gets a target destination and an unordered set of commits to apply. this means it doesnt have the original ordering and thus has to order by ID.

So I'll probably close out for now. its only an issue in this one weird lesson where the goal for one branch is to arbitrarily re-order some commits, which I'll admit almost never has practical application in industry 🙈
